shared/src/dispatchers/ses/sendBulkTemplatedEmail.js

```diff
@@ -44,12 +46,57 @@ exports.sendBulkTemplatedEmail = async ({
       Template: templateName,
     };
 
-    applicationContext.logger.info('Bulk Email Params', params);
+    await exports.sendWithRetry({ applicationContext, params });
+  } catch (err) {
+    applicationContext.logger.error(`Error sending email: ${err}`, err);
+  }
+};
+
+/**
+ * Sends the email via SES with a MAX_SES_RETRIES = 10;
+ *
+ * @param {object} providers the providers object
+ * @param {object} providers.applicationContext application context
+ * @param {object} providers.params the parameters to send to SES
+ * @param {number} providers.retryCount the number of retries attempted
+ */
+exports.sendWithRetry = async ({
+  applicationContext,
+  params,
+  retryCount = 0,
+}) => {
+  const SES = applicationContext.getEmailClient();
+
+  applicationContext.logger.info('Bulk Email Params', params);
+  const response = await SES.sendBulkTemplatedEmail(params).promise();
+  applicationContext.logger.info('Bulk Email Response', response);
 
-    const response = await SES.sendBulkTemplatedEmail(params).promise();
+  // parse response from AWS
+  const needToRetry = response.Status.map((attempt, index) => {
+    // AWS returns 'Success' and helpful identifier upon successful send
+    return attempt.Status !== 'Success' ? params.Destinations[index] : false;
+  }).filter(row => !!row);
 
-    applicationContext.logger.info('Bulk Email Response', response);
-  } catch (err) {
-    applicationContext.logger.error(`Error sending email: ${err.message}`, err);
+  if (needToRetry.length === 0) {
+    return;
+  }
+
+  if (retryCount >= MAX_SES_RETRIES) {
+    const failures = needToRetry
+      .map(dest => dest.Destination.ToAddresses[0])
+      .join(',');
+    throw `Could not complete service to ${failures}`;
   }
+
+  // exponential back-off
+  await backOff(retryCount);
```

Comment:
there is a package called `promise-retry` I have used in the past which has built-in expo backoff which might help reduce the extra logic, but it's probably not critical to use

Reply:
I’ve used the same package in the past as well and have a similar conviction — that is, probably not critical to use, listen to your heart! 🎵